### DISCUSSION TITLE
Duplicate Switched events when using sync jobs

### DISCUSSION THREAD

**Discussion by markkimsal** [tenancy/tenancy#295] [Q&A]
I'm trying to prefix all of my storage drives with a tenant key to add in an extra folder at the root.  This is done by the http ID driver.  But, when I dispatch a sync job, the queue ID driver also kicks off.  It emits a Switched event with the new and old tenants equal to each other.

I can filter out this case in my storage prefix listener, but I'm wondering if there are better solutions.

1. I think that a Switched event shouldn't be submitted at all if the old and new are the same tenant.
2. I think the queue driver should ignore dispatch sync jobs.

For 2. I can't see how a sync job would get dispatched without first going through another ID driver.  Can anybody think of any edge cases for 2?

What about 1? Should equal old and new tenants short circuit the Switched event?

**Comment by thumbtech** [comment]
(You may want to try the support channel in Discord if you don't get a better response here ...)

It seems to work as designed ... does this behavior actually cause an issue/error?

**Reply by markkimsal** [reply]
Yeah kinda, I'm blindly prepending tenant key to storage paths, so I get duplicate folders in my storage whenever I dispatchSync() vs dispatch().

So, if I havea pdf print job, normally would go to job queue with dispatch, then the front-end would poll for the presence of the record. When debugging I use PrintJob::dispatchSync()  so I can wait and see the results.

This leads to saving the pdf into the wrong location because I'm getting 2 switched events.

My problem can be side stepped by not blindly pre-pending.  But, should I really get 2 switched events?

Should dispatchSync re-run queue identification?  I guess I can see the issue if I'm debugging a print job for a tenant that's not already identified by the HTTP ID driver.

But, does the switched event really need to fire when no actual switching happens?  (i.e 2 different ID drivers resolve to the same tenant)

**Reply by markkimsal** [reply]
At the very least it was surprising.  It took a while to figure out what was going on.  _In my mind,_ the queue ID driver was only going to run from the queue and the sync queue driver by-passes the very real db queue I have setup.

I think the queue driver docs could call out that "This runs even with dispatchSync(), so you might end up with Switched events in the middle of your request, not necessarily at the beginning".

Yeah, that's actually really bad now that I think about it.  The remained of the HTTP request would execute under the context of the tenant for the Job.  There's no "switching back".

**Reply by thumbtech** [reply]
I guess the thought process is that the underlying event structure of sync vs other queue drivers should be (almost entirely) the same in order to allow for seamless switching of the drivers.

Probably best to "unblind" the storage path settings for sure!

It is interesting that no "switch back" occurs, although should be fine ... except the edge case would be an override of the queue tenant ... ?

This question was interesting to me because I've evaluated using the queue identifier and I always get cold feet. I always just manually supply tenant model to my jobs and don't use Tenancy there.

**Reply by markkimsal** [reply]
"no switch back should be fine" is the catch 22. If the majority of the time, dispatchSync is done under and already identified tenant, then we don't need to emit Switched for ProcesingJobSync events.  If it is not, then a switch back should occur.

**Reply by ArlonAntonius** [reply]
Hey guys,

Long post incoming, sorry for that 😅 

Completely understand the use case and the Sync Queue driver for Laravel has been bothering me in my head for some time now. As @thumbtech specified; we are trying to keep interoperability between drivers and technically speaking the "sync" queue driver, is still a queue driver. I do agree that this behaviour is weird and something that needs clarification or fixing in 2.x or 3.x ^^.

To answer the two points @markkimsal provided:
1. We could filter that out, but I feel like providing both the oldTenant and the newTenant in the Switching event already allows you to configure this the way that you want. It could use some more DX like a `isSameTenant`, but can also be configured by providing some logic on your side. We're trying to achieve a balance here where everything is possible, but still feels "logical", would love to hear some ideas on how we could make this more friendly without filtering it outright.
2. I think it should be triggered as otherwise we would get quite weird logic on the Sync Queue driver. We would then need a way to allow the developers to specify the drivers that it considers asynchronous queues, which feels like too much work for the user and can easily break with the wrong configuration.

About this not switching back:
This is highly concerning to me and something that we really need to look into. It should 100% switch back as it would otherwise causes unexpected results and state in Tenancy 😅 @markkimsal can you confirm that it never switches back after using a sync job?

**Reply by markkimsal** [reply]
Yeah, I'm only getting 2 Switched events,  one from the HTTP identifier, and one from the Processing event.  But, in my situation, I do not have a setup where the tenant is different from the HTTP identified one and the job.  So not a conclusive answer that it does not "switch back"
